### PR TITLE
Rewrite CodeEditor syntax highlighting for issue #135

### PR DESCRIPTION
**Pull request by erco77**
A rewrite of the style parser to fix issue #135 and to improve the code.
Subject to review, will merge if all is well. Want to see what Travis has to say..

### CODE REVIEW

**Comment by Albrecht-S** [comment]
@erco77 Greg, I'll review the changes when time permits (within the next 2 days or so), I'm pretty busy with other things ATM.

Regarding Travis-CI (hint): you can build your branches in your own fork with Travis-CI (I'm doing this, BTW). I don't know if I recall everything correctly but I believe the _"only thing you need to do"_ ;-) is to **register** with [Travis-CI.com](https://travis-ci.com/) (don't use travis-ci.org, this is going to be closed). You can sign-up with your GitHub account which is straight-forward (IIRC you need to grant travis-ci access to your github account to do its job).

Once this is done, every commit you do in one of your branches with names listed in our `.travis-ci.yml` file will be built automatically and you can view the result. `.travis-ci.yml` contains these valid branch names:
```
branches:
  only:
  - master
  - branch-1.3
  - /^feature-.*/
  - /^fix-.*/
  - /^issue-.*/
  - /^str-.*/
```
Note that all `issue-*` branches are included, but the `'-'` is required. You could also change `.travis-ci.yml` in your branch but I don't recommend this because it would complicate the merge process later (you'd have to reset it at least).

Please let me know if this works for you or not (if you try it at all).

File: fluid/CodeEditor.h [hunk 1]
```diff
@@ -58,6 +55,7 @@ class CodeEditor : public Fl_Text_Editor {
   // attempt to make the fluid code editor widget honour textsize setting
   void textsize(Fl_Fontsize s);
 
+  friend class StyleParse;
```
Comment (Albrecht-S) [inline fluid/CodeEditor.h:58]: I would probably have put the 'friend' statement at the top of the class for better "visibility", but that's a minor issue.

Update copyright year?

File: fluid/StyleParse.cxx [hunk 2]
```diff
@@ -0,0 +1,325 @@
+//
+// Syntax highlighting style parser class - erco 09/16/2020
```
Comment (Albrecht-S) [inline fluid/StyleParse.cxx:2]: Standard header would read "... for the Fast Light Tool Kit (FLTK)" or similar. I'd prefer if you put your "erco ..." in another line (formalism, nitpicking, sorry).

File: fluid/StyleParse.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
+//
+// Syntax highlighting style parser class - erco 09/16/2020
```
Comment (Albrecht-S) [inline fluid/StyleParse.h:2]: "Standard header" comment as above...

File: fluid/StyleParse.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  int  parse_over_angles(char s);
+  int  parse_keyword();           // "switch"
+  int  parse_quoted_string();     // "hello"
+  int  parse_directive();         // "#define"
+  int  parse_line_comment();      // "// text.."
+  int  parse_escape();            // "\'"
+  int  parse_all_else();          // all other code
+};
+
+#endif //StyleParse_h
```
Comment (Albrecht-S) [inline fluid/StyleParse.h:59]: I like `#endif // ...` comments, but I'd suggest a space after `//` as anywhere else. Sorry nitpicking again.

**Review by Albrecht-S — COMMENTED** [review]
So far I only checked the code (diff!) for obvious glitches. Did not test yet or check for (logical) parser issues.

What is the best way to test? Do I need a "complicated" fluid file with C/C++ code, are there any suggestions how to start?

**Comment by erco77** [comment]
> What is the best way to test?

In my case I just made a simple fluid file that had a few instances of New -> Code -> Code, and then open the editor on one of them by double-clicking the code, and then pasting lots of different C and C++ code from other people's C/C++ projects into it, to look for syntax highlighting problems.

File: fluid/StyleParse.cxx (hunk 2 above)
Comment (erco77) [reply fluid/StyleParse.cxx:2]: OK, for consistency with existing headers I can use what we used in Fl_Table/Fl_Tree, which is to use the author's name in the copyright header, e.g.

```
//
// Fl_Table -- A table widget for the Fast Light Tool Kit (FLTK).
//
// Copyright 2002 by Greg Ercolano.
// Copyright (c) 2004 O'ksi'D
//
`

Also, I'll try to remember the "for the Fast Light Tool Kit (FLTK)." suffix; I wasn't aware that was a standard thing, but it seems to be.

File: fluid/StyleParse.h (hunk 4 above)
Comment (erco77) [reply fluid/StyleParse.h]: OK, all the above fixed in da8aeda61

File: fluid/CodeEditor.h (hunk 1 above)
Comment (erco77) [reply fluid/CodeEditor.h]: Done in eea585760. Are we supposed to manage that whenever editing code, or is it something we do en-masse during releases?

**Review by Albrecht-S — APPROVED** [review]
+1

**Comment by erco77** [comment]
Should I take the +1 as approval for a merge, or just the recent updates?
For some reason Travis has a stuck queue ("Some checks haven't completed yet" for a full day now.) Will see if I can trigger a requeue somehow, as I don't want to merge without a check.

**Comment by Albrecht-S** [comment]
> Should I take the +1 as approval for a merge, or just the recent updates?

Oh, sorry for that ambiguity. Basically it was meant for "just the recent updates". However I'm afraid that I won't be able to test soon. If you are confident that the code is fine (works well) feel free to merge. If it's better than before we should go for it.

Regarding Travis CI:
> Guess the 'success' message didn't get from travis -> github.

I've seen that before, seems to happen sometimes.

**Comment by erco77** [comment]
Builds OK on mingw64 and VS2017 (we still have ~400 warnings about precision loss and such).

While pumping a lot of code through the syntax checker, this one thing tripped it up:

	    s.ParseQuotedAtCurPos('"', tmp);

..so I'll provide a fix for that; I wasn't bothering to parse single quotes, but there's the reason to do so, lol.

**Comment by erco77** [comment]
Added single quote handling.
Also, been confirming utf-8 handling throughout, e.g.
unicode-test

**Comment by Albrecht-S** [comment]
> It seems all the .h files in fluid are called .h and not .H. Not sure why.

That's correct, see CMP "[Source File Naming](https://www.fltk.org/cmp.php#CS_SOURCE_FILE_NAMING)":
> ... ".H" for C++ header files in the "FL" directory (".h" otherwise.)

**Pull request merged**